New Astro Runtime Image Doc (#785)
astronomer/docs: astro/runtime-image-architecture.md

---
sidebar_label: 'Astro Runtime Architecture'
title: 'Astro Runtime Architecture'
id: runtime-image-architecture
description: Reference documentation for Astro Runtime, a differentiated distribution of Apache Airflow.
---

Astro Runtime is a production ready, data orchestration tool based on Apache Airflow that is distributed as a Docker image and required by all Astronomer products. It is intended to provide organizations improved functionality, reliability, efficiency, and performance.

## Differences between Astro Runtime and Apache Airflow

If your organization is using Software version 0.29 or later, you can deploy Astro Runtime images. Astro Runtime includes the following features:

> - The `astronomer-providers` package. This package is an open source collection of Apache Airflow providers and modules that is maintained by Astronomer. It includes deferrable versions of popular operators such as `ExternalTaskSensor`, `DatabricksRunNowOperator`, and `SnowflakeOperator`.
> - Airflow UI improvements. For example, showing the Deployment Docker image tag in the footer of all UI pages.
> - Exclusive features such as new Airflow components and improvements to the DAG development experience.
> - OpenLineage with Airflow.  OpenLineage standardizes the definition of data lineage, the metadata that forms lineage data, and how data lineage data is collected from external systems. OpenLineage with Airflow is intended to make it easier for organizations to integrate their lineage tools with Apache Airflow. See [OpenLineage and Airflow](https://docs.astronomer.io/astro/data-lineage-concepts#openlineage-and-airflow).

Reviewer: ```suggestion
- Timely support for new patch, minor, and major versions of Apache Airflow. This includes bug fixes that have not been released in open source but are backported to Astro Runtime.
- Exclusive features to enrich the task execution experience, including smart task concurrency defaults and high availability configurations.
- The `astronomer-providers` package. This package is an open source collection of Apache Airflow providers and modules maintained by Astronomer. It includes deferrable versions of popular operators such as `ExternalTaskSensor`, `DatabricksRunNowOperator`, and `SnowflakeOperator`. See [Astronomer deferrable operators](deferrable-operators.md#astronomer-deferrable-operators).
- The `openlineage-airflow` package. [OpenLineage](https://openlineage.io/) standardizes the definition of data lineage, the metadata that forms lineage data, and how data lineage data is collected from external systems. This package enables data lineage on Astro. See [OpenLineage and Airflow](https://docs.astronomer.io/astro/data-lineage-concepts#openlineage-and-airflow).
- A custom logging module that ensures Airflow task logs are reliably available to the Astro data plane and Astronomer Software.
- A custom security manager that enforces user roles and permissions as defined by Astro and Astronomer Software. See [User permissions](user-permissions.md).
- A custom Airflow UI that includes links to Astronomer resources and exposes the currently running Docker image tag in the footer of all UI pages.
```

Reviewer: Ok @kwcanuck here's my list! Feel free to tweak from here but I think it's a good place to start and covers our bases -- will have to update when new executor comes out

> For more information about the features that are available in Astro Runtime releases, see the [Astro Runtime Release Notes](https://docs.astronomer.io/astro/runtime-release-notes).

Reviewer: ```suggestion
For more information about the features that are available in Astro Runtime releases, see the [Astro Runtime release notes](https://docs.astronomer.io/astro/runtime-release-notes). To install or upgrade Astro Runtime, see [Upgrade Runtime](https://docs.astronomer.io/astro/upgrade-runtime).
```

## Runtime versioning

> Astro Runtime versions are released regularly and use [Semantic Versioning](https://semver.org/). Astronomer ships major, minor, and patch releases of Astro Runtime in the format of `major.minor.patch`.

Reviewer: ```suggestion
Astro Runtime versions are released regularly and use [semantic versioning](https://semver.org/). Astronomer ships major, minor, and patch releases of Astro Runtime in the format of `major.minor.patch`.
```

> - **Major** versions are released for significant feature additions. This includes API or DAG specification changes that are not backwards-compatible.
> - **Minor** versions are released for functional changes. This includes API or DAG specification changes that are backwards-compatible.
> - **Patch** versions are released for bug and security fixes that resolve unwanted behavior. 

Reviewer: ```suggestion
- **Major** versions are released for significant feature additions. This includes new major or minor versions of Apache Airflow as well as API or DAG specification changes that are not backwards-compatible.
- **Minor** versions are released for functional changes. This includes new patch versions of Apache Airflow as well as API or DAG specification changes that are backwards-compatible.
- **Patch** versions are released for bug and security fixes that resolve unwanted behavior. This includes new patch versions of `astronomer-providers` and `openlineage-airflow`.
```

> 
> An **LTS** qualifier identifies Long Term Support versions of Astro Runtime.  

Reviewer: ```suggestion
Every version of Astro Runtime correlates to an Apache Airflow version. All Deployments on Astro must run only one version of Astro Runtime, but you can run different versions of Astro Runtime on different Deployments within a given cluster or Workspace.

For a list of supported Astro Runtime versions and more information on the Astro Runtime maintenance policy, see [Astro Runtime versioning and lifecycle policy](runtime-version-lifecycle-policy.md).
```

- Added some of the versioning-specific language from below up here
- I think let's remove the LTS thing -- the versioning itself won't reflect that status, only our documentation and company policies will. I think just linking to the other doc which is dedicated to describing that in detail is fine?

To install or upgrade Astro Runtime, see [Upgrade Runtime](https://docs.astronomer.io/astro/upgrade-runtime).

## Distribution

> Astro Runtime is distributed as a Debian-based Docker image. The Docker image is functionally identical to open source Apache Airflow. However, Astro Runtime includes performance and stability improvements, including critical bug fixes and security patches.

Reviewer: ```suggestion
Astro Runtime is distributed as a Debian-based Docker image.
```

I think let's remove this -- repetitive with our statements above?

Runtime Docker images have the following format:

- `quay.io/astronomer/astro-runtime:<version>`
- `quay.io/astronomer/astro-runtime:<version>-base`

Astronomer recommends using non-`base` images in your project's `Dockerfile`. These images incorporate Docker ONBUILD commands to copy and scaffold your Astro project directory so you can more easily pass those files to the containers running each core Airflow component. A `base` Astro Runtime image is recommended for complex use cases that require additional customization.

> Astro Runtime Docker images are hosted on the Astronomer Docker registry and enable Airflow on Astro. All Astro projects require that you specify an Astro Runtime image in your Dockerfile, and all Deployments on Astro must run only one version of Astro Runtime. Every version of Astro Runtime correlates to an Apache Airflow version. You can run different versions of Astro Runtime on different Deployments within a given Workspace or Cluster.

Reviewer: ```suggestion
For a list of all Astro Runtime Docker images, see [Quay.io](https://quay.io/repository/astronomer/astro-runtime?tab=tags).

Reviewer: Moved this language up to the versioning section. Also went back and forth on "see this list for all" or just "this is where they're hosted" -- will defer to you.

> The Dockerfiles for all supported Astro Runtime images are located in the [Astronomer `astro-runtime` GitHub repository](https://github.com/astronomer/astro-runtime).

Reviewer: ```suggestion
```

This is a private GH repo, so unfortunately cannot include it

- Amazon [`apache-airflow-providers-amazon`](https://pypi.org/project/apache-airflow-providers-amazon/)
- Elasticsearch [`apache-airflow-providers-elasticsearch`](https://pypi.org/project/apache-airflow-providers-elasticsearch/)
- Celery [`apache-airflow-providers-celery`](https://pypi.org/project/apache-airflow-providers-celery/)
- Google [`apache-airflow-providers-google`](https://pypi.org/project/apache-airflow-providers-google/)
- Password [`apache-airflow-password`](https://pypi.org/project/apache-airflow/)
- Cloud Native Computing Foundation (CNCF) Kubernetes [`apache-airflow-cncf.kubernetes`](https://pypi.org/project/apache-airflow-providers-cncf-kubernetes/)
- PostgreSQL (Postgres) [`apache-airflow-providers-postgres`](https://pypi.org/project/apache-airflow-providers-postgres/)
- Redis [`apache-airflow-providers-redis`](https://pypi.org/project/apache-airflow-providers-redis/)
- StatsD [`apache-airflow-statsd`](https://pypi.org/project/statsd/)
- Virtualenv [`apache-airflow-virtualenv`](https://pypi.org/project/virtualenv/)
- OpenLineage with Airflow [`openlineage-airflow`](https://pypi.org/project/openlineage-airflow/)

> - Astronomer Providers [`astronomer-providers`](https://pypi.org/project/astronomer-providers/)

Reviewer: I think let's add Microsoft -- https://astronomer.slack.com/archives/C01N2DM2ZS8/p1655915600729669

### Provider package versioning

> If an Astro Runtime release includes changes to an installed version of a provider package that is maintained by Astronomer (`astronomer-providers` or `openlineage-airflow`), the version change is documented in the [Astro Runtime Release Notes](/astro/runtime-release-notes.md).

Reviewer: ```suggestion
If an Astro Runtime release includes changes to an installed version of a provider package that is maintained by Astronomer (`astronomer-providers` or `openlineage-airflow`), the version change is documented in the [Astro Runtime release notes](runtime-release-notes.md).
```

To determine the version of a provider package installed in your current Astro Runtime image, run:

```

> docker run --rm {image} pip freeze | grep apache-airflow-provider

Reviewer: ```suggestion
docker run --rm {image} pip freeze | grep <provider>
```

- We typically use `<>` to delineate placeholder values that users should replace with their own
- Not all providers are official Apache Airflow providers, eg. our own so let's keep this general

```

## Python versioning

> Astro Runtime supports Python 3.9. This is the only version of Python that Astro Runtime supports. If your data pipelines require an unsupported Python version, Astronomer recommends that you use the KuberentesPodOperator. See [Run the KubernetesPodOperator on Astro](kubernetespodoperator.md).

Author: @paolaperaza I recommend against adding _currently_. Also, the existing text is clear and concise. There is no ambiguity.

## Astro Runtime and Apache Airflow parity

This table lists Astro Runtime releases and their associated Apache Airflow versions.

| Astro Runtime                                   | Apache Airflow Version                                                                | 
| ----------------------------------------------- | ----------------------------------------------------------------------------- |
| 3.0.x                                       |       2.1.1                                                                       |
| 4.0.x                                       |       2.2.0                                                                    | 
| 4.1.x                                       |       2.2.4                                                                    |
| 4.2.x                                       |       2.2.4                                                                      | 

> | 5.0.x                                       |       2.3.0                                                                        | 

Reviewer: This is what people want to see the most -- let's bring it up in this doc ABOVE provider package section. Maybe we have this as an H3 under `## Versioning`?

Author: @paolaperaza the content has been moved as requested. I would recommend keeping the heading a level 2 as it makes it easier to locate and scan within the topic.

---
sidebar_label: 'Astro Runtime Architecture'
title: 'Astro Runtime Architecture'
id: runtime-image-architecture
description: Reference documentation for Astro Runtime, a differentiated distribution of Apache Airflow.
---

Astro Runtime is a production ready, data orchestration tool based on Apache Airflow that is distributed as a Docker image and required by all Astronomer products. It is intended to provide organizations improved functionality, reliability, efficiency, and performance.

> If your organization is using Software version 0.29 or later, you can deploy Astro Runtime images. Astro Runtime includes the following features:

Reviewer 2: Possible headers:

`## Differences Between Runtime and Apache Airflow`
`## Differences Between Runtime and the Apache Airflow Docker Image`
`## Benefits of Runtime Compared to Apache Airflow`

Author: @jwitz and @paolaperaza I went with _Differences Between Astro Runtime and Apache Airflow_

Reviewer 2: I see this in the Software version of the doc, but not the Astro version.

Author: @jwitz added.

Reviewer: Honestly I'm still wanting to call this `## Features` or `## Benefits` -- I'd rather have the section have a general name to avoid anything that could be interpreted as combative and then in the introductory sentence make it a more explicit comparison.

So:

> ## Features
> 
> If your organization is using Astro, deploying Astro Runtime is required and central to the value of our product. In addition to enabling Apache Airflow on Astronomer, it includes components, configurations, and libraries that enhance performance, usability, and reliability of data orchestration:

Looking at this for inspo: https://docs.databricks.com/runtime/dbr.html. What do you guys think?

- The `astronomer-providers` package. This package is an open source collection of Apache Airflow providers and modules that is maintained by Astronomer. It includes deferrable versions of popular operators such as `ExternalTaskSensor`, `DatabricksRunNowOperator`, and `SnowflakeOperator`.
- Airflow UI improvements. For example, showing the Deployment Docker image tag in the footer of all UI pages.
- Exclusive features such as new Airflow components and improvements to the DAG development experience.

For more information about the features that are available in Astro Runtime releases, see the [Astro Runtime Release Notes](https://docs.astronomer.io/astro/runtime-release-notes).

OpenLineage with Airflow standardizes the definition of data lineage, the metadata that forms lineage data, and how data lineage data is collected from external systems. OpenLineage with Airflow is intended to make it easier for organizations to integrate their lineage tools with Apache Airflow. See [OpenLineage and Airflow](https://docs.astronomer.io/astro/data-lineage-concepts#openlineage-and-airflow).

> The `astronomer-providers` package is a collection of Apache Airflow OSS providers and modules. The package is installed on Astro Runtime by default and is maintained by Astronomer. The package includes deferrable versions of popular operators such as `ExternalTaskSensor`, `DatabricksRunNowOperator`, and `SnowflakeOperator`. See [Deferrable Operators](https://docs.astronomer.io/astro/deferrable-operators). To access the source code for this package, see the [Astronomer Providers GitHub repository](https://github.com/astronomer/astronomer-providers).

Reviewer 2: Besides the doc links, this seems identical to what's in the bulleted list. Let's either:

- Cut these paragraphs entirely and rely on the bulleted list to do the explaining. 
- Very briefly introduce the packages in the bulleted list and expand on them more in these paragraphs.

Author: @jwitz I have removed the paragraph.

- PostgreSQL (Postgres) [`apache-airflow-providers-postgres`](https://pypi.org/project/apache-airflow-providers-postgres/)
- Redis [`apache-airflow-providers-redis`](https://pypi.org/project/apache-airflow-providers-redis/)
- StatsD [`apache-airflow-statsd`](https://pypi.org/project/statsd/)
- Virtualenv [`apache-airflow-virtualenv`](https://pypi.org/project/virtualenv/)
- OpenLineage with Airflow [`openlineage-airflow`](https://pypi.org/project/openlineage-airflow/)
- Astronomer Providers [`astronomer-providers`](https://pypi.org/project/astronomer-providers/)

### Provider Package Versioning

If an Astro Runtime release includes changes to an installed version of a provider package that is maintained by Astronomer (`astronomer-providers` or `openlineage-airflow`), the version change is documented in the [Astro Runtime Release Notes](/astro/runtime-release-notes.md).

To determine the version of a provider package installed in your current Astro Runtime image, run:

```
docker run --rm {image} pip freeze | grep apache-airflow-provider
```

## Python Versioning

> Astro Runtime supports Python package version 3.9.

Reviewer: ```suggestion
Astro Runtime supports Python 3.9.
```

## Astro Runtime and Apache Airflow Parity

This table lists Astro Runtime releases and their associated Apache Airflow versions.

| Astro Runtime                                   | Apache Airflow Version                                                                | 
| ----------------------------------------------- | ----------------------------------------------------------------------------- |
| 3.0.x                                       |       2.1.x                                                                       |
| 4.0.x                                       |       2.2.x                                                                    | 
| 4.1.x                                       |       2.2.x                                                                    |
| 4.2.x                                       |       2.2.x                                                                      | 

> | 5.0.x                                       |       2.3.x                                                                        | 

Reviewer: @kwcanuck Can you add the patch Airflow versions in here? We can keep the `x` for Runtime

Author: @paolaperaza I have returned  the patch Airflow versions.

Reviewer 2: Again, see this in the Software doc but not the Astro doc.

Author: @jwitz updated.

## System Dependencies

The following are the operating system (OS) level dependencies for running basic Astro Runtime system processes:

> - Debian 11.3 or later (bullseye). See [Installing Debian 11.3](https://www.debian.org/releases/bullseye/debian-installer/).

Reviewer: Per conversation in standup, let's do:

- Rename this section from `System Dependencies` to `System Distribution`
- Simplify our statement here to, `Astro Runtime images are based on Debian 11.3 (Bullseye)`...

Author: @paolaperaza implemented as requested.

- A custom Airflow UI that includes links to Astronomer resources and exposes the currently running Docker image tag in the footer of all UI pages.

For more information about the features that are available in Astro Runtime releases, see the [Astro Runtime release notes](runtime-release-notes.md). To install or upgrade Astro Runtime, see [Upgrade Runtime](upgrade-runtime.md).

## Runtime versioning

Astro Runtime versions are released regularly and use [semantic versioning](https://semver.org/). Astronomer ships major, minor, and patch releases of Astro Runtime in the format of `major.minor.patch`.

- **Major** versions are released for significant feature additions. This includes new major or minor versions of Apache Airflow as well as API or DAG specification changes that are not backwards-compatible.
- **Minor** versions are released for functional changes. This includes new patch versions of Apache Airflow as well as API or DAG specification changes that are backwards-compatible.
- **Patch** versions are released for bug and security fixes that resolve unwanted behavior. This includes new patch versions of `astronomer-providers` and `openlineage-airflow`.

> Every version of Astro Runtime correlates to an Apache Airflow version. All Deployments on Astro must run only one version of Astro Runtime, but you can run different versions of Astro Runtime on different Deployments within a given cluster or Workspace.

Reviewer: ```suggestion
Every version of Astro Runtime correlates to an Apache Airflow version. All Deployments on Astro must run only one version of Astro Runtime, but you can run different versions of Astro Runtime on different Deployments within a given cluster or Workspace. See [Create a Deployment](create-deployment.md#create-a-deployment).
```

For a list of supported Astro Runtime versions and more information on the Astro Runtime maintenance policy, see [Astro Runtime versioning and lifecycle policy](runtime-version-lifecycle-policy.md).

> To install or upgrade Astro Runtime, see [Upgrade Runtime](https://docs.astronomer.io/astro/upgrade-runtime).

Reviewer: This could go below table?

## Distribution

Astro Runtime is distributed as a Debian-based Docker image. Runtime Docker images have the following format:

- `quay.io/astronomer/astro-runtime:<version>`
- `quay.io/astronomer/astro-runtime:<version>-base`

An Astro Runtime image must be specified in the `Dockerfile` of your Astro project. Astronomer recommends using non-`base` images, which incorporate ONBUILD commands that copy and scaffold your Astro project directory so you can more easily pass those files to the containers running each core Airflow component. A `base` Astro Runtime image is recommended for complex use cases that require additional customization, such as [installing Python packages from private sources](develop-project.md#install-python-packages-from-private-sources).

For a list of all Astro Runtime Docker images, see [Quay.io](https://quay.io/repository/astronomer/astro-runtime?tab=tags).

## Executors

In Airflow, the executor is responsible for determining how and where a task is completed.

> In all local environments created with the Astro CLI, Astro Runtime runs the [Local executor](https://airflow.apache.org/docs/apache-airflow/stable/executor/local.html).
> 
> On Astro, Astro Runtime exclusively supports the [Celery executor](https://airflow.apache.org/docs/apache-airflow/stable/executor/celery.html).

Reviewer: ```suggestion
In all local environments created with the Astro CLI, Astro Runtime runs the [Local executor](https://airflow.apache.org/docs/apache-airflow/stable/executor/local.html). On Astro, Astro Runtime exclusively supports the [Celery executor](https://airflow.apache.org/docs/apache-airflow/stable/executor/celery.html).
```

Soon, Astronomer will provide a new executor with intelligent worker packing, task-level resource requests, improved logging, and Kubernetes-like task isolation.

> ## Provider packages

Reviewer: Move this as an H2 above, below `### Astro Runtime and Apache Airflow parity` and above `### Provider versioning`? I see that you've grouped all the "versioning" topics together, but I suppose it's a little weird that we talk about provider versioning before we even define providers.
